diff --git a/Build/BundledPrograms.hs b/Build/BundledPrograms.hs
index 096eeee29..cca18c899 100644
--- a/Build/BundledPrograms.hs
+++ b/Build/BundledPrograms.hs
@@ -87,3 +87,10 @@ preferredBundledPrograms = catMaybes
 	ifset True s = Just s
 	ifset False _ = Nothing
 #endif
+
+magicBundle :: [FilePath]
+#ifdef mingw32_HOST_OS
+magicBundle = ["libmagic-1.dll", "libgnurx-0.dll", "magic.mgc"]
+#else
+magicBundle = []
+#endif
diff --git a/Build/NullSoftInstaller.hs b/Build/NullSoftInstaller.hs
index 6d8ca3a29..95da8e4d8 100644
--- a/Build/NullSoftInstaller.hs
+++ b/Build/NullSoftInstaller.hs
@@ -42,6 +42,7 @@ main = do
 	withTmpDir "nsis-build" $ \tmpdir -> do
 		let gitannex = tmpdir </> gitannexprogram
 		mustSucceed "ln" [File "git-annex.exe", File gitannex]
+                mapM_ (\f -> mustSucceed "ln" [File f, File (tmpdir </> f)]) magicBundle
 		let license = tmpdir </> licensefile
 		mustSucceed "sh" [Param "-c", Param $ "zcat standalone/licences.gz > '" ++ license ++ "'"]
 		webappscript <- vbsLauncher tmpdir "git-annex-webapp" "git annex webapp"
@@ -51,7 +52,7 @@ main = do
 		let gitannexcmd = tmpdir </> "git-annex.cmd"
 		writeFile gitannexcmd "git annex %*"
 		writeFile nsifile $ makeInstaller
-			gitannex gitannexcmd license htmlhelp winPrograms
+			gitannex gitannexcmd license htmlhelp (winPrograms ++ magicBundle)
 			[ webappscript, autostartscript ]
 		mustSucceed "makensis" [File nsifile]
 	removeFile nsifile -- left behind if makensis fails
